File: ios/App/App/Discover/UDPDiscoveryPlugin.m

//
//  CapacitorView.m
//  App
//
//  Created by Max Lynch on 6/5/19.
//

#import <Foundation/Foundation.h>

#import <Capacitor/Capacitor.h>

CAP_PLUGIN(UDPDiscovery, "UDPDiscovery",
           CAP_PLUGIN_METHOD(start, CAPPluginReturnPromise);
           CAP_PLUGIN_METHOD(stop, CAPPluginReturnPromise);
           CAP_PLUGIN_METHOD(getServices, CAPPluginReturnPromise);
           )

File: ios/App/App/Discover/UDPDiscoveryPlugin.swift
import Foundation
import Capacitor

@objc(UDPDiscovery)
public class UDPDiscovery : CAPPlugin {
  
  var service: UDPDiscoveryService!
  var callbackID: String?
  
  public override func load() {
    service = UDPDiscoveryService(namespace: "devapp")
  }
  
  @objc func start(_ call: CAPPluginCall) {
    service.close()
    service.start { }
  }
  
  @objc func stop(_ call: CAPPluginCall) {
    service.close()
    call.resolve()
  }
  
  @objc func getServices(_ call: CAPPluginCall) {
    let message = generateMessage(service.getServices())
    call.resolve(message)
  }
  
  private func generateMessage(_ services: [Service]) -> [String: Any] {
    let arrayOfServices = services.map { (s) -> [String: Any] in
      return s.toDictionary()
    }
    return ["services": arrayOfServices]
  }
}
  
